lib/matplotlib/mpl-data/matplotlibrc

```diff
@@ -433,6 +433,10 @@
 #axes3d.yaxis.panecolor:    (0.90, 0.90, 0.90, 0.5)  # background pane on 3D axes
 #axes3d.zaxis.panecolor:    (0.925, 0.925, 0.925, 0.5)  # background pane on 3D axes
 
+#axes3d.mouserotationstyle: arcball  # {azel, trackball, arcball, Shoemake, Holroyd}
```

Comment:
IMO `'Holroyd'` should be the default over `'arcball'`

Reply:
...but `'Holroyd'` is weird in several respects: 
- the control surface (a hyperbola) has no obvious mechanical equivalent; it is more of an *ad hoc* solution
- the 'roll' angular speed does not match the angular speed of the mouse when near the edge
- if you do several rotations in succession in different directions, but come back to the same point on screen with the mouse, then the orientation turns out unchanged, which is unnatural.

Taking all of this into account, it seems Holroyd is altogether not very unsuitable for a first encounter: 
the 'principle of least surprise' gets short shrift.
Some of this applies to the Shoemake style as well. I'd argue that `arcball` is the most innocent of the options.

Reply:
As matplotlib doesn't support physical trackballs at the moment, I don't see any value in anchoring against mechanical analogues when more abstracted methods provide desirable properties. I feel very strongly that the lack of hysteresis is a good property to have for a mouse-based interface, since it lets people easily "undo" errant clicks. 

> if you do several rotations in succession in different directions, but come back to the same point on screen with the mouse, then the orientation turns out unchanged; there might be advantageous not to have hysteresis, but very natural it is not

Several rotations will not keep this property! Rather, the lack of hysteresis is a property within a _single_ mouse click. If I put my finger on a ball and roll it around while keeping contact, I would expect the ball to return to its original orientation when my finger does.

Reply:
> the 'roll' angular speed does not match the angular speed of the mouse when near the edge

This I'm not so convinced is an issue, but it does bring up that the Holroyd method does not let you control pure roll like some of the other methods, since its control surface never reaches 0. I can't actually find the original Holroyd method on the referenced website in the paper (https://members.tripod.com/professor_tom/), but it definitely appears ad-hoc (frankly, it just seems like the author _really likes_ hyperbolic surfaces).

But what if we make a new method that recovers this property? I would propose "filleting" the edges of the Shoemake control surface hemisphere with radius 1 circular curvature, like this in `_arcball`. This smooths out the control response like Holroyd's method, but also allows you to control purely roll on the edges:

```python
        elif style == 'matplotlib':
            r = np.sqrt(r2)
            if r > np.sqrt(3):
                p = np.array([0, x, y])
            elif r > np.sqrt(3)/2:
                p = np.array([1 - np.sqrt(1 - (np.sqrt(3)-r)**2), x, y])
            else:
                p = np.array([math.sqrt(1-r2), x, y])
```

![image](https://github.com/user-attachments/assets/908a635c-863d-41de-9df7-fa718270e212)

Reply:
Playing around with this method feels very nice. It also helped me understand why the "double speed" rotations are actually a good thing, and necessary for path independence. For example, if you are dragging from the top of the screen to the bottom here, you can take a pure roll path or a pure elevation path. For those two paths to reach the same orientation, the view has to do a 360 deg flip, not a 180 deg flip.

![image](https://github.com/user-attachments/assets/a8ab9cf0-a3ed-470c-a084-d613a5952e3b)

Reply:
Equations come from this geometry of the cross section of the surface.

![image](https://github.com/user-attachments/assets/1d5b0735-dd32-42c4-b0e0-6172ce493c50)

Reply:
> > the 'roll' angular speed does not match the angular speed of the mouse when near the edge
> 
> The speed I'm not so convinced is an issue, [...]
Well, it is a bit a matter of taste, I suppose - what may be less of an issue to you could be more of an issue to someone else...
To me, it is somewhat disconcerting that after moving the mouse around the perimeter by just half a circle, that the figure already has made a full turn. (Commonly, a mouse is use for 'dragging' something - not for making it hurry ahead.)
But if that does not bother you, and you'd happily accept that, in return for something that is more important, more desirable to you, like ending up with the same rotation independent of the path taken by the mouse, then adjusting the rcparam setting could be a good solution, I thought. To my understanding, the settings are for preferences; if there would be a solution where you could have both the natural angular speed and the path independence, I'd be all for it, but I'm not aware of any such method - so it's a choice, hence the rcParam... Does that make sense?

Reply:
> [...] but it does illuminate a shortcoming of the Holroyd method that we weren't thinking about before - it does not let you control pure roll like some of the other methods, since its control surface never reaches 0. I can't actually find the original Holroyd method on the referenced website in the paper (https://members.tripod.com/professor_tom/), but it definitely appears ad-hoc (frankly, the choice of a hyperbolic roll-off just seems like it's because the author _really likes_ hyperbolic surfaces).

:hand_over_mouth:
> But what if we make a new method that recovers this property? I would propose "filleting" the edges of the Shoemake hemispherical control surface with radius 1 circular curvature, which could be accomplished by adding the below to `_arcball`. This smooths out the control response like Holroyd's method, and also allows you to control purely roll on the edges:
> 
> ```python
>         elif style == 'matplotlib':
>             r = np.sqrt(r2)
>             if r > np.sqrt(3):
>                 p = np.array([0, x, y])
>             elif r > np.sqrt(3)/2:
>                 p = np.array([1 - np.sqrt(1 - (np.sqrt(3)-r)**2), x, y])
>             else:
>                 p = np.array([math.sqrt(1-r2), x, y])
> ```
Well I think there is something to be said for that, I like the idea, I wouldn't be opposed to including a 'Shambaugh' mouse rotation style like that. One concern that I have is that the "filetting" border is rather wide, and one wouldn't really be able to get to the outer region for reasonable settings of the ''trackballsize'' (that is, 0.66*sqrt(3) = 1.14... is beyond the edge of the axes (at ±1); only in the corners you'd have a chance...)

Reply:
> Playing around with this method feels very nice. It also helped me understand why the "double speed" rotations are necessary for path independence. As an illuminating example, if you are dragging from the top of the screen to the bottom here, you can take a pure roll path or a pure elevation path. For those two paths to reach the same orientation, the view must do a 360 deg flip, not a 180 deg flip.

Yes, yes, you got it ! - the "double speed" / "rotation is independent of path taken" is a choice: you can fix the former, at the price of sacrificing the latter. The question is - what is preferable? Is there a unequivocal answer to that?

Reply:
> To clarify what I think you're (correctly) saying - [...] If I put my finger on a ball and roll it around while keeping contact, I would expect the ball to return to its original orientation when my finger does. I think the mechanical analogue of a finger-on-a-ball is the right one to use when thinking about mouse interactions - it certainly feels more natural to me and for users with touchscreens it will be literally true.

I think the finger-on-a-ball is not the right mechanical analogue. You cannot `torque' with a mouse: after you rotate your mouse in-place (or your finger on a touchscreen), the mouse arrow will still point in the same direction (usually to the top left, pretty much forever). Indeed, if you could torque with a mouse, then there would be no _need_ for a trackball: one could just use the mouse torque for the roll, and be done. 
Since the mouse cursor is apparently just a point then, we need the trackball, and we need to somehow make a sensible connection between mouse movements and the change in orientation of the ball. The 'arcball' style amounts to touching a ball, torque free. But the 'Shoemake' arcball... what does it correspond to...?

Reply:
> If I put my finger on a ball and roll it around while keeping contact, I would expect the ball to return to its original orientation when my finger does.

If you would do this _without torquing_, then the ball would not quite return to its original orientation, in general... that's the point.

Reply:
If you make the radius of the fillet `a = 0.25`, it fits rather nicely in the view volume, with a bit of pure roll control surface around the edge. Here's what that control surface looks like when plotted:

![image](https://github.com/user-attachments/assets/c61c8d8e-2042-4ff7-afa9-f5f3eab62ec5)

Here are the equations: 
![image](https://github.com/user-attachments/assets/b95c0711-5c77-45d8-aecb-adaa93793f8a)

And here's the code:
```python
        elif style == 'matplotlib':
            r = np.sqrt(r2)
            a = 0.25
            ra = np.sqrt(1 + 2*a)
            ri = np.sqrt(1 + 2*a)/(1 + a)
            if r < ri:
                p = np.array([np.sqrt(1 - r**2), x, y])
            elif r < ra:
                p = np.array([a - np.sqrt(a**2 - (r - np.sqrt(1 + 2*a))**2), x, y])
            else:
                p = np.array([0, x, y])
```

Reply:
> Well I think there is something to be said for that, I like the idea, I wouldn't be opposed to including a 'Shambaugh' mouse rotation style like that.

I'm glad you like it! And while I'm flattered, it's a collaborative effort so I'd recommend calling it 'matplotlib' or something generic. I do think it's superior to the Bell method because of the pure roll control zone. 

And thinking about it, you could flatten the top to have a pure azim/el zone in the center of your screen... but perhaps that's too much to bring in all at once.

Reply:
I'm not quite following the "torquing" line of thinking. I have a physical ball on my desk which I can roll around with my finger and have it return to the original position every time without lifting my finger. But it's a bit of a moot point, let's continue discussion in the main comments

Reply:
> I'm not quite following the "torquing" line of thinking. 

Oh I should explain it better, then -
> I have a physical ball on my desk which I can roll around with my finger and have it return to the original position every time without lifting my finger.

... the situation with a mouse is more like rolling the physical ball around using the point of a sharp pencil, instead of a finger. You can turn that pencil along its axis, but nothing really happens then to the physical ball; just like you can twist your mouse in-place, but nothing really happens to the mouse cursor.

> But it's a bit of a moot point, let's continue discussion in the main comments

Oh I thought it was the essence. The point is, turning a mouse on its back does not turn it into a trackball - one degree of freedom is missing (you cannot apply 'roll' to the ball in those old mechanical mice, only in fancy trackball mice you can do that).

Reply:
> Here is the geometry: ![image](https://private-user-images.githubusercontent.com/14363975/374003906-b95c0711-5c77-45d8-aecb-adaa93793f8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgzNDcxMTAsIm5iZiI6MTcyODM0NjgxMCwicGF0aCI6Ii8xNDM2Mzk3NS8zNzQwMDM5MDYtYjk1YzA3MTEtNWM3Ny00NWQ4LWFlY2ItYWRhYTkzNzkzZjhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA4VDAwMjAxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYjkxOWViMzA2OTA5YjQxMDdiMWRmZjdlNzY2NDNlZTc5NzNkNmE3ZjE4M2IxOTQ3MjczMGVlYmQzMzIzNTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qlpva0y4wOYo0PizC6mxm27Z27aNqI22bErjuE5R01U)
> 
> And here's the code:
> 
> ```python
>         elif style == 'matplotlib':
>             r = np.sqrt(r2)
>             a = 0.25
>             ra = np.sqrt(1 + 2*a)
>             ri = np.sqrt(1 + 2*a)/(1 + a)
>             if r < ri:
>                 p = np.array([np.sqrt(1 - r**2), x, y])
>             elif r < ra:
>                 p = np.array([a - np.sqrt(a**2 - (r - np.sqrt(1 + 2*a))**2), x, y])
>             else:
>                 p = np.array([0, x, y])
> ```

I thought it might be nice to parametrize the width of the border a bit differently: rather than using $a$, we could use the projection along the surface (horizontally in the picture). It is more directly connected to what's going on on screen, and simplifies the math a bit.  Let's say $b$ is the extra space that the new border will take up beyond where Shoemake's edge originally was, then $r_a = 1+b$. Inverting $r_a=\sqrt{1+2a}$, we obtain $a = b(1+ \frac b 2)$.
Instead of $r_i = \frac{r_a }{1 + a}$, we could consider doing $r_i = 2 /(r_a + 1/r_a)$. That takes one extra division, but has as advantage that it works not only for $b \rightarrow 0$ but also for $b = \infty$.
We can also rewrite $z = a - \sqrt{a^2 - (r - r_a)^2}$ a bit; introducing $dr = r_a - r$, it becomes
 $z = a - \sqrt{(a+dr)(a-dr)}$, which avoids the small difference between two large squares. Unfortunately, it still does not work well for very large $b$ (and hence even larger $a$), it evaluates to $0$, instead of $1$ (the resulting user experience is not good). (I tried to think of a fix, using $\sin(dr/a)$, but that then messes up the $b \rightarrow 0$, $a \rightarrow 0$ limit, which is more important to me; the `b=1#inf` case reacts a bit like a trackball, but not quite: it slows down away from the center, so it is not a replacement for the 'trackball' style, anyway).
One thing that still needs fixing above is that the vector resulting from _`arcball()` is supposed to be on the unit sphere; currently it isn't.

The good thing with the limit $b \rightarrow 0$ is that it recovers the Shoemake arcball, so we do not need to have a separate 'matplotlib' arcball, we can consolidate to just 4: 'azel', 'trackball', 'arcball', and 'Shoemake'. I'm tempted to rename the latter, and in the spirit of Shoemake's paper name it 'ARCBALL' (with capitals). Both 'arcball' and 'ARCBALL' then have the optional border.
Does that make sense?

Reply:
Your new parameterization makes sense to me, let's do it. I don't think we need to really worry about the $b \rightarrow \inf$ case.

I don't think that we should rely on case sensitivity to distinguish different configurations, as that could be a big source of confusion. 

The naming is a little confusing to me in general, because the literature calls Shoemake's approach arcball. Is what is currently labeled 'arcball' described elsewhere? The sqrt compared to the Shoemake arcball make it seem novel to me but I might be missing a source. IMO the cleanest naming would be to make the changes `'Shoemake' -> 'arcball'`, `'arcball' -> something else`. I'm not too hung up the particular here though.

Reply:
> I don't think that we should rely on case sensitivity to distinguish different configurations, as that could be a big source of confusion.
> 
> The naming is a little confusing to me in general, because the literature calls Shoemake's approach arcball. Is what is currently labeled 'arcball' described elsewhere? The sqrt compared to the Shoemake arcball make it seem novel to me but I might be missing a source. Separately, 'Shoemake' as a name makes it seem like we are referring to a very particular control scheme, rather than a generalization of his idea. IMO the cleanest naming would be to make the changes `'Shoemake' -> 'arcball'` (I previously proposed `'matplotlib'` here, can keep or drop), and `'arcball' -> something else`. I'm not too hung up the particulars here though, your call.

Yes, let's rename:
- 'Shoemake' -> 'arcball', since that's the name that Shoemake gave to it, and since we want to avoid giving the impression that we refer to a very particular control scheme (we tinkered with the border, after all)
- what was 'arcball' before -> 'sphere', simply.

I'll push a commit, shortly.